packages/store-sync/src/watchLogs.ts

```diff
@@ -23,137 +26,227 @@ type WatchLogsEvent = {
+          const blocks = groupLogsByBlockNumber(logs);
+          debug("releasing", logs.length, "logs across", blocks.length, "blocks to subscriber");
+          for (const block of blocks) {
+            subscriber.next(block);
+          }
+          caughtUp = true;
+          logBuffer.length = 0;
+        })
+        .catch((error) => {
+          const errorMessage = `ws${wsId} failed to fetch initial logs`;
+          debugError(errorMessage, error);
+          subscriber.error(errorMessage);
+        });
     }
 
-    setupClient().catch((error) => {
-      debug("error setting up initial client", error);
-      subscriber.error(error);
+    setup().catch((error) => {
+      const errorMessage = `ws${wsId} failed to setup wiresaw_watchLogs subscription`;
+      debugError(errorMessage, error);
+      subscriber.error(errorMessage);
     });
 
+    // Send a ping to keep the connection alive
+    const ping = setInterval(
+      () =>
+        ws &&
+        request({ ws, method: "net_version", wsId }).catch(() => {
+          const errorMessage = `ws${wsId} ping failed`;
+          debugError(errorMessage);
+          subscriber.error(errorMessage);
+        }),
+      10_000,
+    );
+
     return () => {
-      debug("logs$ subscription closed, closing client");
+      debug(`ws${wsId} logs$ subscription closed, closing client`);
+      clearInterval(ping);
       try {
-        client?.close();
-        client?.socket?.close();
+        ws?.close();
       } catch (e) {
-        debug("failed to close client/socket", e);
+        debug(`ws${wsId} failed to close web socket`, e);
       }
     };
   });
 
   return { logs$ };
 }
 
-type FetchInitialLogsInput = { client: SocketRpcClient<WebSocket>; topics: LogTopic[] } & Omit<WatchLogsInput, "url">;
+async function waitForWebSocketOpen(ws: WebSocket, wsId?: number): Promise<void> {
+  if (ws.readyState !== WebSocket.OPEN) {
+    const [resolve, reject, promise] = deferred<void>();
+    debug(`ws${wsId} waiting for websocket to open`);
+    const timeout = setTimeout(() => reject(new Error(`ws${wsId} timeout waiting for websocket to open`)), 10_000);
+    ws.addEventListener("open", () => {
+      clearTimeout(timeout);
+      resolve();
+    });
+    await promise;
+  }
+}
+
+type RequestAsyncArgs = {
+  ws: WebSocket;
+  method: string;
+  params?: unknown[];
+  wsId: number;
+};
+
+async function request<T>({ ws, method, params, wsId }: RequestAsyncArgs): Promise<T> {
+  await waitForWebSocketOpen(ws, wsId);
+  const requestId = uuid();
+  const [resolve, reject, promise] = deferred<T>();
+
+  debug(`ws${wsId} sending request`, method, requestId);
+  ws.send(JSON.stringify({ jsonrpc: "2.0", id: requestId, method, params }), (error) => {
+    if (error) {
+      debugError(`ws${wsId} request error`, error);
+      reject(error);
+    }
+  });
+
+  const timeout = setTimeout(() => reject(new Error("timeout waiting for response for " + requestId)), 10_000);
```

Comment:
should this remove the message listener?

Reply:
this will cause the connection to be closed and recreated so don't necessarily have to remove the msg listener on the old ws